clash-lib/src/Clash/Normalize/Util.hs

```diff
@@ -423,30 +425,67 @@ normalizeTopLvlBndr isTop nm (nm',sp,inl,tm) = makeCachedU nm (extra.normalized)
 --
 -- we create the substitution [dom := "System"] and apply it to create:
 --
--- > \(eta : Signal "System" Bool) . eta
+-- > \(eq : "System" ~ "System") . \(eta : Signal "System" Bool) . eta
 --
 -- __NB:__ Users of this function should ensure it's only applied to TopEntities
 substWithTyEq
-  :: [TyVar]
-  -> [(TyVar,Type)]
-  -> [Id]
-  -> Term
+  :: Term
   -> Maybe Term
   -- ^ 'Nothing' if 'substWithTyEq' didn't have to substitute anything
-substWithTyEq tvs cvs ids_ (TyLam tv e) = substWithTyEq (tv:tvs) cvs ids_ e
-substWithTyEq tvs cvs ids_ (Lam v e)
-  | TyConApp (nameUniq -> tcUniq) [_,VarTy tv, ty] <- tyView (varType v)
-  , tcUniq == getKey eqTyConKey
-  , tv `elem` tvs
-  = substWithTyEq (tvs List.\\ [tv]) ((tv,ty):cvs) (v:ids_) e
-substWithTyEq tvs cvs@(_:_) ids_ e =
-  let
-    e1 = List.foldl' (flip TyLam) e tvs
-    subst0 = extendTvSubstList (mkSubst emptyInScopeSet) cvs
-    subst1 = extendIdSubstList subst0 [(v, removedTm (varType v)) | v <- ids_]
-  in
-    Just (substTm "substWithTyEq" subst1 e1)
-substWithTyEq _ _ _ _ = Nothing
+substWithTyEq = go [] False []
+ where
+  go
+    :: [TyVar]
+    -> Bool
+    -> [Id]
+    -> Term
+    -> Maybe Term
+  go tvs changed ids_ (TyLam tv e) = go (tv:tvs) changed ids_ e
+  go tvs changed ids_ (Lam v e)
+    | TyConApp (nameUniq -> tcUniq) (tvFirst -> Just (tv, ty)) <- tyView (varType v)
+    , tcUniq == getKey eqTyConKey
+    , tv `elem` tvs
+    = let
+        subst0 = extendTvSubst (mkSubst emptyInScopeSet) tv ty
+        subst1 = extendIdSubst subst0 v (removedTm (varType v))
+      in go (tvs List.\\ [tv]) True (substId subst0 v : ids_) (substTm "substWithTyEq e" subst1 e)
+    | otherwise = go tvs changed (v:ids_) e
+  go tvs True ids_ e =
+    let
+      e1 = List.foldl' (flip TyLam) e tvs
+      e2 = List.foldl' (flip Lam) e1 ids_
+    in
+      Just e2
+  go _ _ _ _ = Nothing
+
+-- Type equality (~) is symmetrical, so users could write: (dom ~ System) or (System ~ dom)
+tvFirst :: [Type] -> Maybe (TyVar, Type)
+tvFirst [_, VarTy tv, ty] = Just (tv, ty)
+tvFirst [_, ty, VarTy tv] = Just (tv, ty)
+tvFirst _ = Nothing
+
+-- | The type equivalent of 'substWithTyEq'
+tvSubstWithTyEq
+  :: Type
+  -> {- Maybe -} Type
```

Comment:
I had the Maybe there in comments because `substWithTyEq` also has a Maybe result, and `tvSubstWithTyEq` is basically the same as `substWithTyEq` but for `Type`s.

I've removed the Maybe because I found it just annoying to use.
It now behaves the same(returning the original type) as if it would be called like:
```
 Maybe.fromMaybe ty0 (tvSubstWithTyEq ty0)
```